import numpy, seaborn


import pandas
pandas.set_option('display.max_colwidth', None)


import scipy, scipy.signal, scipy.optimize
import statsmodels, statsmodels.stats, statsmodels.stats.multitest


import matplotlib, matplotlib.pyplot
matplotlib.rcParams.update({'font.size':20, 
                            'font.family':'sans-serif', 
                            'xtick.labelsize':16, 
                            'ytick.labelsize':16, 
                            'figure.figsize':(16*(2/3), 9*(2/3)), 
                            'axes.labelsize':20
                           })





# read biomarkers
df = pandas.read_excel('/Users/adrian/hub/vigur/002.transcriptomics/arnar_biomarkers/Endothelialbiomarkers.xlsx')
print(df.shape)
biomarkers = list(set(df['Gene']))
print(len(biomarkers))

for element in biomarkers:
    the_list = list(df['Gene'])
    counter = the_list.count(element)
    if counter > 1:
        print(f'Element {element} is twice')
print(len(biomarkers), biomarkers[:5])


# read dose responders
df = pandas.read_csv('/Users/adrian/hub/vigur/003.proteomics/dose_responders_gene_results.txt', sep='\s+', index_col=0)
print(df.shape)

together = list(set(biomarkers + list(set(df['gene_name']))))
print(len(together), together)

biomarkers = together

## SELE is the only overlap btw biomarkers and dose responders, isnt that interesting!





# read colored proteins 04 
df = pandas.read_csv('../colored_proteins_04.tsv', sep='\t')
print(df.columns)
df = df[df["Color_in_plot"].isin(["red", "blue"])]
intersect = list(set(biomarkers) & set(df['Gene_name']))
sub = df[df['Gene_name'].isin(intersect)]
hits04 = sub['Uniprot_gene_ID'].to_list()
print(len(hits04), hits04)
sub


# read colored proteins 24 
df = pandas.read_csv('../colored_proteins_24.tsv', sep='\t')
print(df.columns)
df = df[df["Color_in_plot"].isin(["red", "blue"])]
intersect = list(set(biomarkers) & set(df['Gene_name']))
sub = df[df['Gene_name'].isin(intersect)]
hits24 = sub['Uniprot_gene_ID'].to_list()
print(len(hits24), hits24)
sub


hits = list(set(hits04 + hits24))
print(len(hits))
print(hits)





input_file = '../230117_Arnar.lfq.proteins_From_Gudrun.xlsx'
df = pandas.read_excel(input_file)
df.head()





# drop C9JKI3_HUMAN bc is the same as CAV1 but poorer quality
print(hits, len(hits))
hits.remove('C9JKI3_HUMAN')
print(len(hits))

# Build a regex pattern like "Q01105|Q9H2X3|Q99873"
pattern = "|".join(hits)
print('biomarkers length', len(hits), hits[:3])

# Subset
df = df[df["Accession"].str.contains(pattern, na=False)]
df

### why there is not perfect match!??!?


gene_names = []
for element in df['Description']:
    gene_name = element.split('GN=')[1].split(' ')[0]
    gene_names.append(gene_name)

print(gene_names)
print(len(gene_names))
print(len(set(gene_names)))


df['gene_names'] = gene_names
df = df.set_index("gene_names")


selected_columns = {}
working_columns = []

selected_columns['CTRL_4H_'] = []
selected_columns['CTRL_24H_'] = []
selected_columns['ADRNOR_4H_'] = []
selected_columns['ADRNOR_24H_'] = []

for element in df.columns:
    for key in selected_columns.keys():
        if key in element:
            selected_columns[key].append(element)
            working_columns.append(element)

for key in selected_columns.keys():
    print(key)
    print(selected_columns[key])
    print()
    
print(working_columns)


df = df[working_columns]
df


# lets get the medians of the injections. well, there are two injections, median, mean, potato, potato...
next_columns = []
for i in range(0, len(working_columns), 2):
    col_a = working_columns[i]
    col_b = working_columns[i + 1]
    new_col_name = 'median_' + '_'.join(col_a.split('_')[:3]) 
    df[new_col_name] = df[[col_a, col_b]].median(axis=1)
    print(col_a, col_b, new_col_name)
    next_columns.append(new_col_name)

df = df[next_columns]
df


# lets get the median over biological replicates
working_columns = df.columns
next_columns = []
for i in range(0, len(working_columns), 3):
    col_a = working_columns[i]
    col_b = working_columns[i + 1]
    col_c = working_columns[i + 2]
    new_col_name = 'median_' + '_'.join(col_a.split('_')[1:3]) 
    df[new_col_name] = df[[col_a, col_b, col_c]].median(axis=1)
    print(col_a, col_b, col_c, new_col_name)
    next_columns.append(new_col_name)

df = df[next_columns]
df





zscore_df = df.apply(scipy.stats.zscore, axis=1)
zscore_df.head()


new_order = ["median_CTRL_4H", "median_ADRNOR_4H", "median_CTRL_24H", "median_ADRNOR_24H"]
zscore_df = zscore_df[new_order]


labels04 = [element.split('_')[0] for element in hits04]
labels24 = [element.split('_')[0] for element in hits24]




print(labels04)
col_4h = 'median_ADRNOR_4H'
replacements = {
    'TRBM': 'THBD',
    'LAT1': 'SLC7A5',
    'PECA1': 'PECAM1',
    'CTR2': 'SLC7A2',
}

labels04 = [replacements.get(x, x) for x in labels04]


annot = pandas.DataFrame(
    '',
    index=zscore_df.index,
    columns=zscore_df.columns,
    dtype=object
)
for gene in labels04:
    if gene in annot.index:
        annot.at[gene, col_4h] = '*'
    else:
        print('bypassing', gene)



replacements = {
    
    'IL8': 'CXCL8',          # interleukin-8 is now CXCL8
    'TM87A': 'TMEM87A',      # alias TM87A → TMEM87A
    'E9PKG1': 'DDAH2',       # UniProt E9PKG1 maps to DDAH2
    'ALBU': 'ALB',           # albumin → ALB
    'PGH2': 'PTGS2',         # prostaglandin G/H synthase 2 → PTGS2
}


zscore_df


linkage_method = 'average'
distance_metric = 'cosine'

seaborn.set(font_scale=0.9)

seaborn.clustermap(zscore_df, 
                   annot=annot, fmt='', annot_kws={'va':'center_baseline', 'size':18},
                   cmap='bwr', 
                   col_cluster=False, 
                   row_cluster=True,
                   vmin=-2.5, vmax=2.5, 
                   method=linkage_method, 
                   metric=distance_metric, 
                   yticklabels=1, 
                   cbar_kws={'label':'z-score'})

matplotlib.pyplot.tight_layout()

matplotlib.pyplot.savefig('friday.svg')






